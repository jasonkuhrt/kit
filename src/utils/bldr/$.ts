export * as Bldr from './$$.js'
