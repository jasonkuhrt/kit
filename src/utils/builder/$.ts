export * as Builder from './$$.js'
