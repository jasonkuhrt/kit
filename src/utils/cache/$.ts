export * as Cache from './$$.ts'
