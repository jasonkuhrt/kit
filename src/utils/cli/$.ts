export * as Cli from './$$.js'
