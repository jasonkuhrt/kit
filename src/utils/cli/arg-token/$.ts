export { ArgToken } from './arg-token.js'
