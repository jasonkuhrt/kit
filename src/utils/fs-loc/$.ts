export * as FsLoc from './$$.js'
