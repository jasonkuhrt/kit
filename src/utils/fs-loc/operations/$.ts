// Re-export all operations

export * from './ensure-absolute.js'
export * from './ensure-optional-absolute-with-cwd.js'
export * from './ensure-optional-absolute.js'
export * from './extension.js'
export * from './is-above.js'
export * from './is-root.js'
export * from './is-under.js'
export * from './join.js'
export * from './name.js'
export * from './stem.js'
export * from './to-abs.js'
export * from './to-dir.js'
export * from './to-rel.js'
export * from './up.js'
