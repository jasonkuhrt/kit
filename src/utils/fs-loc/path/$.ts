export * as Path from './$$.js'
