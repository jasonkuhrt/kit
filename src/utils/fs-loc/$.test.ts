import { Test } from '#test'
import '../test/matchers/$.js'
import { Schema as S } from 'effect'
import { describe, expect, expectTypeOf, it } from 'vitest'
import { FsLoc } from './$.js'
import './$.test-matchers.js'

const P = FsLoc.Path
const F = FsLoc.File
const l = FsLoc.fromString

// dprint-ignore
Test.describe('.decodeSync')
  .on(S.decodeSync(FsLoc.FsLoc))
  .describe('AbsFile')
  .casesAsArg(
    '/file.txt',
    '/home/user/doc.pdf',
    '/a/b/c/d/e.js',
    '/archive.tar.gz',
    '/.config.json',
    '/my docs/file name.txt',
  )
  .describe('RelFile')
  .casesAsArg(
    'file.txt',
    './file.txt',
    '../file.txt',
    'src/index.ts',
    '../../lib/util.js',
    './src/components/App.tsx',
  )
  .describe('AbsDir')
  .casesAsArg(
    '/',
    '/home/',
    '/home',
    '/usr/local/bin/',
    '/a/b/c/d/e/',
    '/my documents/projects/',
  )
  .describe('RelDir')
  .casesAsArg(
    './',
    'src/',
    '../',
    'src/components/',
    '../../lib/',
    './src/',
    '../src/lib/utils/',
    'src',
  )
  .test()

const LocLoose = FsLoc.FsLocLoose.make
// dprint-ignore
Test.describe('.FsLocLoose.decodeSync')
  .i<string>()
  .o<FsLoc.FsLocLoose.FsLocLooseClass>()
  .cases(
    ['abs file',      ['/home/file.txt'],           LocLoose({ path: P.Abs.make({ segments: ['home'] }), file: F.make({ name: 'file', extension: '.txt' }) })],
    ['abs dir',       ['/home/'],                   LocLoose({ path: P.Abs.make({ segments: ['home'] }), file: null })],
    ['rel file',      ['file.txt'],                 LocLoose({ path: P.Rel.make({ segments: [] }), file: F.make({ name: 'file', extension: '.txt' }) })],
    ['rel dir',       ['src/'],                     LocLoose({ path: P.Rel.make({ segments: ['src'] }), file: null })],
    ['root',          ['/'],                        LocLoose({ path: P.Abs.make({ segments: [] }), file: null })],
    // Files without extensions are treated as directories - changed to .js extension
    ['complex path',  ['/usr/local/bin/node.js'],   LocLoose({ path: P.Abs.make({ segments: ['usr', 'local', 'bin'] }), file: F.make({ name: 'node', extension: '.js' }) })],
  )
  .test((i: string, o: any) => {
    const result = FsLoc.FsLocLoose.decodeSync(i)
    expect(result).toBeEquivalent(o, FsLoc.FsLocLoose.FsLocLooseClass)
  })

describe('.Constants', () => {
  describe('absDirRoot', () => {
    it('represents the root directory /', () => {
      expect(FsLoc.Constants.absDirRoot).toEqual(l('/'))
      expect(S.encodeSync(FsLoc.AbsDir.String as any)(FsLoc.Constants.absDirRoot)).toBe('/')
    })
  })

  describe('relDirCurrent', () => {
    it('represents the current directory .', () => {
      expect(FsLoc.Constants.relDirCurrent).toEqual(l('./'))
      expect(S.encodeSync(FsLoc.RelDir.String as any)(FsLoc.Constants.relDirCurrent)).toBe('./')
    })
  })

  describe('relDirParent', () => {
    it('represents the parent directory ..', () => {
      expect(FsLoc.Constants.relDirParent).toEqual(l('../'))
      expect(S.encodeSync(FsLoc.RelDir.String as any)(FsLoc.Constants.relDirParent)).toBe('./../')
    })
  })
})

describe('.fromString', () => {
  it('returns AbsFile for absolute file paths', () => {
    const result = FsLoc.fromString('/path/file.txt')
    expect(result._tag).toBe('LocAbsFile')
    expectTypeOf(result).toEqualTypeOf<FsLoc.AbsFile>()
  })

  it('returns RelFile for relative file paths', () => {
    const result = FsLoc.fromString('file.txt')
    expect(result._tag).toBe('LocRelFile')
    expectTypeOf(result).toEqualTypeOf<FsLoc.RelFile>()
  })

  it('returns AbsDir for absolute directory paths', () => {
    const result = FsLoc.fromString('/home/')
    expect(result._tag).toBe('LocAbsDir')
    expectTypeOf(result).toEqualTypeOf<FsLoc.AbsDir>()
  })

  it('returns RelDir for relative directory paths', () => {
    const result = FsLoc.fromString('./src/')
    expect(result._tag).toBe('LocRelDir')
    expectTypeOf(result).toEqualTypeOf<FsLoc.RelDir>()
  })

  it('type-errors when given non-literal string', () => {
    // @ts-expect-error: fromString requires a literal string
    FsLoc.fromString('/' as string)
  })
})

// dprint-ignore
Test.describe('Groups *.assert')
  .i<string>()
  .o<{ assert: Function; pass: boolean }>()
  .cases(
    ['Rel passes for file.txt',      ['file.txt'],  { assert: FsLoc.Groups.Rel.assert,  pass: true }],
    ['Rel fails for /file.txt',      ['/file.txt'], { assert: FsLoc.Groups.Rel.assert,  pass: false }],

    ['Abs passes for /file.txt',     ['/file.txt'], { assert: FsLoc.Groups.Abs.assert,  pass: true }],
    ['Abs fails for file.txt',       ['file.txt'],  { assert: FsLoc.Groups.Abs.assert,  pass: false }],

    ['File passes for file.txt',     ['file.txt'],  { assert: FsLoc.Groups.File.assert, pass: true }],
    ['File fails for ./src/',        ['./src/'],    { assert: FsLoc.Groups.File.assert, pass: false }],

    ['Dir passes for ./src/',        ['./src/'],    { assert: FsLoc.Groups.Dir.assert,  pass: true }],
    ['Dir fails for file.txt',       ['file.txt'],  { assert: FsLoc.Groups.Dir.assert,  pass: false }],

    ['AbsFile passes for /file.txt',  ['/file.txt'],  { assert: S.asserts(FsLoc.AbsFile),  pass: true }],
    ['AbsFile fails for file.txt',    ['file.txt'],   { assert: S.asserts(FsLoc.AbsFile),  pass: false }],
    ['AbsFile fails for /dir/',       ['/dir/'],      { assert: S.asserts(FsLoc.AbsFile),  pass: false }],

    ['RelFile passes for file.txt',   ['file.txt'],   { assert: S.asserts(FsLoc.RelFile),  pass: true }],
    ['RelFile fails for /file.txt',   ['/file.txt'],  { assert: S.asserts(FsLoc.RelFile),  pass: false }],
    ['RelFile fails for ./dir/',      ['./dir/'],     { assert: S.asserts(FsLoc.RelFile),  pass: false }],

    ['AbsDir passes for /dir/',       ['/dir/'],      { assert: S.asserts(FsLoc.AbsDir),   pass: true }],
    ['AbsDir fails for ./dir/',       ['./dir/'],     { assert: S.asserts(FsLoc.AbsDir),   pass: false }],
    ['AbsDir fails for /file.txt',    ['/file.txt'],  { assert: S.asserts(FsLoc.AbsDir),   pass: false }],

    ['RelDir passes for ./dir/',      ['./dir/'],     { assert: S.asserts(FsLoc.RelDir),   pass: true }],
    ['RelDir fails for /dir/',        ['/dir/'],      { assert: S.asserts(FsLoc.RelDir),   pass: false }],
    ['RelDir fails for file.txt',     ['file.txt'],   { assert: S.asserts(FsLoc.RelDir),   pass: false }],
  )
  .test((i: string, o: any) => {
    const loc = FsLoc.decodeSync(i)
    if (o.pass) {
      expect(() => o.assert(loc)).not.toThrow()
    } else {
      expect(() => o.assert(loc)).toThrow()
    }
  })
