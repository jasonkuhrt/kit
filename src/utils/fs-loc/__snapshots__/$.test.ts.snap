// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AbsDir > .decodeSync > fn("/") > fn("/") 1`] = `
AbsDir {
  "_tag": "LocAbsDir",
  "path": Abs {
    "_tag": "PathAbs",
    "segments": [],
  },
}
`;

exports[`AbsDir > .decodeSync > fn("/a/b/c/d/e/") > fn("/a/b/c/d/e/") 1`] = `
AbsDir {
  "_tag": "LocAbsDir",
  "path": Abs {
    "_tag": "PathAbs",
    "segments": [
      "a",
      "b",
      "c",
      "d",
      "e",
    ],
  },
}
`;

exports[`AbsDir > .decodeSync > fn("/home") > fn("/home") 1`] = `
AbsDir {
  "_tag": "LocAbsDir",
  "path": Abs {
    "_tag": "PathAbs",
    "segments": [
      "home",
    ],
  },
}
`;

exports[`AbsDir > .decodeSync > fn("/home/") > fn("/home/") 1`] = `
AbsDir {
  "_tag": "LocAbsDir",
  "path": Abs {
    "_tag": "PathAbs",
    "segments": [
      "home",
    ],
  },
}
`;

exports[`AbsDir > .decodeSync > fn("/my documents/projects/") > fn("/my documents/projects/") 1`] = `
AbsDir {
  "_tag": "LocAbsDir",
  "path": Abs {
    "_tag": "PathAbs",
    "segments": [
      "my documents",
      "projects",
    ],
  },
}
`;

exports[`AbsDir > .decodeSync > fn("/usr/local/bin/") > fn("/usr/local/bin/") 1`] = `
AbsDir {
  "_tag": "LocAbsDir",
  "path": Abs {
    "_tag": "PathAbs",
    "segments": [
      "usr",
      "local",
      "bin",
    ],
  },
}
`;

exports[`AbsFile > .decodeSync > fn("/.config.json") > fn("/.config.json") 1`] = `
AbsFile {
  "_tag": "LocAbsFile",
  "file": File {
    "_tag": "File",
    "extension": ".json",
    "name": ".config",
  },
  "path": Abs {
    "_tag": "PathAbs",
    "segments": [],
  },
}
`;

exports[`AbsFile > .decodeSync > fn("/a/b/c/d/e.js") > fn("/a/b/c/d/e.js") 1`] = `
AbsFile {
  "_tag": "LocAbsFile",
  "file": File {
    "_tag": "File",
    "extension": ".js",
    "name": "e",
  },
  "path": Abs {
    "_tag": "PathAbs",
    "segments": [
      "a",
      "b",
      "c",
      "d",
    ],
  },
}
`;

exports[`AbsFile > .decodeSync > fn("/archive.tar.gz") > fn("/archive.tar.gz") 1`] = `
AbsFile {
  "_tag": "LocAbsFile",
  "file": File {
    "_tag": "File",
    "extension": ".gz",
    "name": "archive.tar",
  },
  "path": Abs {
    "_tag": "PathAbs",
    "segments": [],
  },
}
`;

exports[`AbsFile > .decodeSync > fn("/file.txt") > fn("/file.txt") 1`] = `
AbsFile {
  "_tag": "LocAbsFile",
  "file": File {
    "_tag": "File",
    "extension": ".txt",
    "name": "file",
  },
  "path": Abs {
    "_tag": "PathAbs",
    "segments": [],
  },
}
`;

exports[`AbsFile > .decodeSync > fn("/home/user/doc.pdf") > fn("/home/user/doc.pdf") 1`] = `
AbsFile {
  "_tag": "LocAbsFile",
  "file": File {
    "_tag": "File",
    "extension": ".pdf",
    "name": "doc",
  },
  "path": Abs {
    "_tag": "PathAbs",
    "segments": [
      "home",
      "user",
    ],
  },
}
`;

exports[`AbsFile > .decodeSync > fn("/my docs/file name.txt") > fn("/my docs/file name.txt") 1`] = `
AbsFile {
  "_tag": "LocAbsFile",
  "file": File {
    "_tag": "File",
    "extension": ".txt",
    "name": "file name",
  },
  "path": Abs {
    "_tag": "PathAbs",
    "segments": [
      "my docs",
    ],
  },
}
`;

exports[`RelDir > .decodeSync > fn("../") > fn("../") 1`] = `
RelDir {
  "_tag": "LocRelDir",
  "path": Rel {
    "_tag": "PathRelative",
    "segments": [
      "..",
    ],
  },
}
`;

exports[`RelDir > .decodeSync > fn("../../lib/") > fn("../../lib/") 1`] = `
RelDir {
  "_tag": "LocRelDir",
  "path": Rel {
    "_tag": "PathRelative",
    "segments": [
      "..",
      "..",
      "lib",
    ],
  },
}
`;

exports[`RelDir > .decodeSync > fn("../src/lib/utils/") > fn("../src/lib/utils/") 1`] = `
RelDir {
  "_tag": "LocRelDir",
  "path": Rel {
    "_tag": "PathRelative",
    "segments": [
      "..",
      "src",
      "lib",
      "utils",
    ],
  },
}
`;

exports[`RelDir > .decodeSync > fn("./") > fn("./") 1`] = `
RelDir {
  "_tag": "LocRelDir",
  "path": Rel {
    "_tag": "PathRelative",
    "segments": [],
  },
}
`;

exports[`RelDir > .decodeSync > fn("./src/") > fn("./src/") 1`] = `
RelDir {
  "_tag": "LocRelDir",
  "path": Rel {
    "_tag": "PathRelative",
    "segments": [
      "src",
    ],
  },
}
`;

exports[`RelDir > .decodeSync > fn("src") > fn("src") 1`] = `
RelDir {
  "_tag": "LocRelDir",
  "path": Rel {
    "_tag": "PathRelative",
    "segments": [
      "src",
    ],
  },
}
`;

exports[`RelDir > .decodeSync > fn("src/") > fn("src/") 1`] = `
RelDir {
  "_tag": "LocRelDir",
  "path": Rel {
    "_tag": "PathRelative",
    "segments": [
      "src",
    ],
  },
}
`;

exports[`RelDir > .decodeSync > fn("src/components/") > fn("src/components/") 1`] = `
RelDir {
  "_tag": "LocRelDir",
  "path": Rel {
    "_tag": "PathRelative",
    "segments": [
      "src",
      "components",
    ],
  },
}
`;

exports[`RelFile > .decodeSync > fn("../../lib/util.js") > fn("../../lib/util.js") 1`] = `
RelFile {
  "_tag": "LocRelFile",
  "file": File {
    "_tag": "File",
    "extension": ".js",
    "name": "util",
  },
  "path": Rel {
    "_tag": "PathRelative",
    "segments": [
      "..",
      "..",
      "lib",
    ],
  },
}
`;

exports[`RelFile > .decodeSync > fn("../file.txt") > fn("../file.txt") 1`] = `
RelFile {
  "_tag": "LocRelFile",
  "file": File {
    "_tag": "File",
    "extension": ".txt",
    "name": "file",
  },
  "path": Rel {
    "_tag": "PathRelative",
    "segments": [
      "..",
    ],
  },
}
`;

exports[`RelFile > .decodeSync > fn("./file.txt") > fn("./file.txt") 1`] = `
RelFile {
  "_tag": "LocRelFile",
  "file": File {
    "_tag": "File",
    "extension": ".txt",
    "name": "file",
  },
  "path": Rel {
    "_tag": "PathRelative",
    "segments": [],
  },
}
`;

exports[`RelFile > .decodeSync > fn("./src/components/App.tsx") > fn("./src/components/App.tsx") 1`] = `
RelFile {
  "_tag": "LocRelFile",
  "file": File {
    "_tag": "File",
    "extension": ".tsx",
    "name": "App",
  },
  "path": Rel {
    "_tag": "PathRelative",
    "segments": [
      "src",
      "components",
    ],
  },
}
`;

exports[`RelFile > .decodeSync > fn("file.txt") > fn("file.txt") 1`] = `
RelFile {
  "_tag": "LocRelFile",
  "file": File {
    "_tag": "File",
    "extension": ".txt",
    "name": "file",
  },
  "path": Rel {
    "_tag": "PathRelative",
    "segments": [],
  },
}
`;

exports[`RelFile > .decodeSync > fn("src/index.ts") > fn("src/index.ts") 1`] = `
RelFile {
  "_tag": "LocRelFile",
  "file": File {
    "_tag": "File",
    "extension": ".ts",
    "name": "index",
  },
  "path": Rel {
    "_tag": "PathRelative",
    "segments": [
      "src",
    ],
  },
}
`;
